server/src/main/java/org/elasticsearch/cluster/metadata/TemplateUpgradeService.java

```diff
@@ -151,6 +151,9 @@ void updateTemplates(Map<String, BytesReference> changes, Set<String> deletions)
             client.admin().indices().putTemplate(request, new ActionListener<PutIndexTemplateResponse>() {
                 @Override
                 public void onResponse(PutIndexTemplateResponse response) {
+                    if (calculateTemplateChanges(clusterService.state().getMetaData().getTemplates()).isPresent()) {
```

Comment:
this check needs to be tightened up as it can fail when there are multiple template updates. The change should probably be something like:

```java
Optional<Tuple<Map<String, BytesReference>, Set<String>>> changes = calculateTemplateChanges(clusterService.state().getMetaData().getTemplates());
if (changes.isPresent() && changes.get().v1().contains(request.name())) {
    throw new IllegalStateException("...");
}
```

Also, can you add a unit test for this scenario?